src/toast/ops/totalconvolve.py

```diff
@@ -0,0 +1,695 @@
+            # Loop over views
+            views = obs.view[self.view]
+            for view in range(len(views)):
+                # Get the flags if needed
+                flags = None
+                if self.apply_flags:
+                    if self.shared_flags is not None:
+                        flags = np.array(views.shared[self.shared_flags][view])
+                        flags &= self.shared_flag_mask
+                    if self.det_flags is not None:
+                        detflags = np.array(views.detdata[self.det_flags][view][det])
+                        detflags &= self.det_flag_mask
+                        if flags is not None:
+                            flags |= detflags
+                        else:
+                            flags = detflags
+
+                # Timestream of detector quaternions
+                quats = views.detdata[self.detector_pointing.quats][view][det]
+                if verbose:
+                    timer.report_clear(f"get detector pointing for {det}")
+
+                if flags is not None:
+                    quats = quats.copy()
+                    quats[flags != 0] = nullquat
+                    if verbose:
+                        timer.report_clear(f"initialize flags for detector {det}")
+
+                theta, phi, psi = qa.to_angles(quats)
+                # Polarization angle in the Pxx basis
+                psi_pol = self._get_psi_pol(focalplane, det)
+                if self.dxx:
+                    # Add angle between Dxx and Pxx
+                    psi_pol += self._get_psi_uv(focalplane, det)
+                psi -= psi_pol
+                psi_pol = np.ones(psi.size) * psi_pol
+                if self.hwp_angle is not None:
+                    hwp_angle = views.shared[self.hwp_angle][view]
+                    psi_pol += 2 * hwp_angle
+                all_theta.append(theta)
+                all_phi.append(phi)
+                all_psi.append(psi)
+                all_psi_pol.append(psi_pol)
+
+        if len(all_theta) > 0:
+            all_theta = np.hstack(all_theta)
+            all_phi = np.hstack(all_phi)
+            all_psi = np.hstack(all_psi)
+            all_psi_pol = np.hstack(all_psi_pol)
+
+        if verbose:
+            timer.report_clear(f"compute pointing angles for detector {det}")
+        return all_theta, all_phi, all_psi, all_psi_pol
+
+    def get_buffer(self, theta, phi, psi, det, verbose):
+        """Pack the pointing into the pointing array"""
+        timer = Timer()
+        timer.start()
+        pnt = np.empty((len(theta), 3))
+        pnt[:, 0] = theta
+        pnt[:, 1] = phi
+        pnt[:, 2] = psi + np.pi  # FIXME: not clear yet why this is necessary
+        if verbose:
+            timer.report_clear(f"pack input array for detector {det}")
+        return pnt
+
+    def convolve(self, sky, beam, lmax, mmax, pnt, det, verbose):
+        timer = Timer()
+        timer.start()
+        convolver = totalconvolve.Interpolator(
```

Comment:
I guess here is where we need to do something different if we have an MPI communicator?  In that case, every process in `self.comm` will have some amount of pointing data (possibly zero samples, if we are working on a detector that does not exist on every process).  And also every process will be doing a portion of the convolution.

Alternatively, if implementing the communication needed for that is too much at the moment, it would also be fine to put a check in the `_exec()` method and raise an exception if `self.comm.size` is greater than one.  This would allow us to run certain types of jobs for testing in the mean time.

Reply:
If we want to be efficient, we ned to do something more clever in the presence of MPI, but the code will also work (somewhat wastefully) as it is. Every task will do the full convolution work then and store the full data cube. The interpolation part will be parallelized automatically, I guess, due to the pointings being distributed over the tasks already. In short, this simple version will be sort of OK as long as `lmax` and `beammmax` are not very high, and you are processing very many pointings with every call.

The ultimate solution would be to use the "ConvolverPlan" class from `ducc` instead of te simpler `Interpolator` class. Then the approach would be:
- distribute the convolution work among the tasks along the m_beam axis
- redistribute the data such that every task has the full m_beam axis, but only parts of theta
- FFT the m_beam axis into a psi axis
- rearrange the pointings such that every task only has pointings falling into its own theta range
- do the interpolation
- move the resulting data back to its original tasks

However, this is a bit more than I have time for at the moment ...